community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/TransactionStream.java

```diff
@@ -0,0 +1,341 @@
+
+package org.neo4j.kernel.impl.store.counts;
+
+import java.io.File;
+import java.io.FilenameFilter;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.neo4j.io.fs.FileSystemAbstraction;
+import org.neo4j.io.fs.StoreChannel;
+import org.neo4j.kernel.impl.transaction.CommittedTransactionRepresentation;
+import org.neo4j.kernel.impl.transaction.command.Command;
+import org.neo4j.kernel.impl.transaction.log.IOCursor;
+import org.neo4j.kernel.impl.transaction.log.LogDeserializer;
+import org.neo4j.kernel.impl.transaction.log.PhysicalLogFile;
+import org.neo4j.kernel.impl.transaction.log.PhysicalLogVersionedStoreChannel;
+import org.neo4j.kernel.impl.transaction.log.PhysicalTransactionRepresentation;
+import org.neo4j.kernel.impl.transaction.log.ReadAheadLogChannel;
+import org.neo4j.kernel.impl.transaction.log.entry.LogEntry;
+import org.neo4j.kernel.impl.transaction.log.entry.LogEntryCommand;
+import org.neo4j.kernel.impl.transaction.log.entry.LogEntryCommit;
+import org.neo4j.kernel.impl.transaction.log.entry.LogEntryStart;
+import org.neo4j.kernel.impl.transaction.log.entry.LogHeader;
+
+import static javax.transaction.xa.Xid.MAXBQUALSIZE;
+import static javax.transaction.xa.Xid.MAXGTRIDSIZE;
+
+import static org.neo4j.kernel.impl.transaction.log.LogVersionBridge.NO_MORE_CHANNELS;
+import static org.neo4j.kernel.impl.transaction.log.ReadAheadLogChannel.DEFAULT_READ_AHEAD_SIZE;
+import static org.neo4j.kernel.impl.transaction.log.entry.LogHeaderReader.readLogHeader;
+
+class TransactionStream
+{
+    private final LogFile[] files;
+    private final FileSystemAbstraction fs;
+    private final boolean isContiguous;
+    private final LogFile firstFile;
+
+    public TransactionStream( FileSystemAbstraction fs, File path ) throws IOException
+    {
+        this.fs = fs;
+        { // read metadata from all files in the directory
+            ByteBuffer buffer = ByteBuffer.allocateDirect( 9 + MAXGTRIDSIZE + MAXBQUALSIZE * 10 );
+            File[] files = fs.listFiles( path, new TxFileFilter( fs ) );
+            ArrayList<LogFile> logFiles = new ArrayList<>( files.length );
+            for ( File file : files )
+            {
+                LogFile logFile = new LogFile( fs, file, buffer );
+                if ( logFile.lastCommittedTxId > 0 )
+                {
+                    logFiles.add( logFile );
+                }
+            }
+            this.files = logFiles.toArray( new LogFile[logFiles.size()] );
+        }
+        Arrays.sort( files );
+        { // verify the read metadata
+            LogFile first = null, last = null;
+            boolean isContiguous = true;
+            for ( LogFile file : files )
+            {
+                if ( last == null )
+                {
+                    first = file;
+                }
+                else if ( isContiguous && last.lastCommittedTxId < file.header.lastCommittedTxId )
```

Comment:
I believe you are wrong. On the left hand side of the `<` we have the last committed transaction in the previous file we looked at, and on the right hand side we have the `lastCommittedTxId` field from the header of the current file, i.e. what that file believes was the transaction committed in the file before it.

If file A spans `]7,83]` and file B spans `]83,122]` we would be comparing `A.lastTx < B.header.lastTx` ⇔ `83 < 83` ⇔ `false`, and would conclude that these two files are contiguous.

If file A spans `]7,83]` and file B spans `]84,122]` we would be comparing `A.lastTx < B.header.lastTx` ⇔ `83 < 84` ⇔ `true` and would conclude that these two files are *not* contiguous.

If file A spans `]7,83]` and file B spans `]76,122]` we would be comparing `A.lastTx < B.header.lastTx` ⇔ `83 < 76` ⇔ `false` and would conclude that these two files are contiguous. This comparison allows for overlap between  the files, but does not tolerate gaps.

Please explain to me where my reasoning is wrong.